clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java

```diff
@@ -36,76 +36,101 @@
-        bufferOffset = 0;
-        bufferSize = 0;
-        finished = false;
+        this.checkHC = false;
+        this.headerRead = false;
+        this.finished = false;
+    }
+
+    /**
+     * Idempotent read of header/FrameDescriptor
+     */
+    private void ensureHeaderRead() throws IOException {
+        if (!this.headerRead) {
+            readHeader();
+            this.headerRead = true;
+        }
     }
 
     /**
      * Reads the magic number and frame descriptor from the underlying {@link InputStream}.
      *
      * @throws IOException
      */
-    private void readHeader() throws IOException {
+    private void readHeader() throws IOException, LZ4Exception {
         byte[] header = new byte[LZ4_MAX_HEADER_LENGTH];
 
         // read first 6 bytes into buffer to check magic and FLG/BD descriptor flags
-        bufferOffset = 6;
-        if (in.read(header, 0, bufferOffset) != bufferOffset) {
+        int headerOffset = 6;
+        if (in.read(header, 0, headerOffset) != headerOffset) {
             throw new IOException(PREMATURE_EOS);
         }
 
-        if (MAGIC != Utils.readUnsignedIntLE(header, bufferOffset - 6)) {
-            throw new IOException(NOT_SUPPORTED);
+        if (MAGIC != Utils.readUnsignedIntLE(header, headerOffset - 6)) {
+            throw new LZ4Exception(NOT_SUPPORTED);
         }
-        flg = FLG.fromByte(header[bufferOffset - 2]);
-        bd = BD.fromByte(header[bufferOffset - 1]);
-        // TODO read uncompressed content size, update flg.validate()
-        // TODO read dictionary id, update flg.validate()
+        this.flg = FLG.fromByte(header[headerOffset - 2]);
+        this.bd = BD.fromByte(header[headerOffset - 1]);
+
+        this.maxBlockSize = bd.getBlockMaximumSize();
+        this.compressedBuffer = new byte[maxBlockSize];
+        this.buffer = new byte[maxBlockSize];
+        this.bufferOffset = 0;
+        this.bufferSize = 0;
 
-        // check stream descriptor hash
-        byte hash = (byte) ((checksum.hash(header, 0, bufferOffset, 0) >> 8) & 0xFF);
-        header[bufferOffset++] = (byte) in.read();
-        if (hash != header[bufferOffset - 1]) {
-            throw new IOException(DESCRIPTOR_HASH_MISMATCH);
+        if (this.flg.isContentSizeSet()) {
+            if (in.read(header, headerOffset, 8) != 8)
+                throw new IOException(PREMATURE_EOS);
+            headerOffset += 8;
         }
+
+        // Final byte of Frame Descriptor is HC checksum
+        header[headerOffset++] = (byte) in.read();
+
+        // Old implementations produced incorrect HC checksums
+        if (!this.checkHC)
+            return;
+
+        int offset = 4;
+        int len = headerOffset - offset - 1; // dont include magic bytes or HC
+        assert len >= 2 && len <= 10 : "Checksum: Invalid FrameDescriptor size";
+        byte hash = (byte) ((checksum.hash(header, offset, len, 0) >> 8) & 0xFF);
+        if (hash != header[headerOffset - 1])
```

Comment:
My understanding is that Jun is suggesting that we should set `checkHC` to true for the Java client if message format is 1.